daisi/src/cpps/logical/order_management/simple_order_management.cpp

```diff
@@ -0,0 +1,199 @@
+  }
+  const auto orders = task.getOrders();
+  auto order_it = std::find(orders.rbegin(), orders.rend(), order);
+  const int order_index = orders.rend() - order_it - 1;
+
+  // determine end location of previous order
+  std::optional<Location> previous_location;
+  order_it++;
+  for (; order_it != orders.rend(); order_it++) {
+    previous_location = getEndLocationOfOrder(*order_it);
+    if (previous_location.has_value()) {
+      break;
+    }
+  }
+  daisi::util::Position previous_position;
+  if (!previous_location.has_value()) {
+    if (!current_end_position_.has_value()) {
+      // previous_position = current_pose_.position_;
+      throw std::logic_error("there must exist a location for the amr to start from");
+    }
+    // else {
+    previous_position = current_end_position_.value();
+    // }
+  } else {
+    previous_position = previous_location.value().getPosition();
+  }
+  auto funcs = materialFlowToFunctionalities({order}, previous_position);
+
+  if (std::holds_alternative<MoveOrder>(order)) {
+    if (order_index == 0) {
+      throw std::invalid_argument("move order should not be first");
+    }
+    metrics.empty_travel_time += AmrMobilityHelper::estimateDuration(
+        daisi::util::Pose{}, funcs, amr_description_, topology_, false);
+
+    metrics.empty_travel_distance = AmrMobilityHelper::calculateDistance(previous_position, funcs);
+  } else if (std::holds_alternative<TransportOrder>(order)) {
+    auto res = AmrMobilityHelper::calculateMetricsByDomain(previous_position, funcs,
+                                                           amr_description_, topology_);
+    metrics.empty_travel_time += std::get<0>(res);
+    metrics.loaded_travel_time += std::get<1>(res);
+    metrics.action_time += std::get<2>(res);
+    metrics.empty_travel_distance += std::get<3>(res);
+    metrics.loaded_travel_distance += std::get<4>(res);
+  } else if (std::holds_alternative<ActionOrder>(order)) {
+    metrics.action_time += AmrMobilityHelper::estimateDuration(daisi::util::Pose{}, funcs,
+                                                               amr_description_, topology_, false);
+  } else {
+    throw std::invalid_argument("Order type not supported");
+  }
+}
+
+std::optional<Location> SimpleOrderManagement::getEndLocationOfOrder(const Order &order) {
+  return std::visit(
+      [&](auto &&arg) -> std::optional<Location> {
+        using T = std::decay_t<decltype(arg)>;
+        if constexpr (std::is_same_v<T, MoveOrder>) {
+          return std::get<MoveOrder>(order).getMoveOrderStep().getLocation();
+        } else if constexpr (std::is_same_v<T, TransportOrder>) {
+          return std::get<TransportOrder>(order).getDeliveryTransportOrderStep().getLocation();
+        } else if constexpr (std::is_same_v<T, ActionOrder>) {
+          return std::nullopt;
+        } else {
+          static_assert(always_false_v<T>, "Order type not handled");
+        }
+      },
+      order);
+}
```

Comment:
I thought about this duplication of code. 

We can turn this into a static method because it does not require any information of the `SimpleOrderManagement` or `StnOrderManagement` class. 

Let's create a `order_management_helper.h`and `order_management_helper.cpp` for such static helper methods.